worlds/dark_souls_3/__init__.py

```diff
@@ -185,364 +223,1350 @@ def create_connection(from_region: str, to_region: str):
+            return
+
+        location = self.random.choice(candidate_locations)
+        location.place_locked_item(item)
+
+    def _replace_with_filler(self, location: DarkSouls3Location) -> None:
+        """If possible, choose a filler item to replace location's current contents with."""
+        if location.locked: return
+
+        # Try 10 filler items. If none of them work, give up and leave it as-is.
+        for _ in range(0, 10):
+            candidate = self.create_filler()
+            if location.item_rule(candidate):
+                location.item = candidate
+                return
 
     def get_filler_item_name(self) -> str:
-        return "Soul of an Intrepid Hero"
-
+        return self.random.choice(filler_item_names)
 
     def set_rules(self) -> None:
-        # Define the access rules to the entrances
-        set_rule(self.multiworld.get_entrance("Go To Undead Settlement", self.player),
-                 lambda state: state.has("Small Lothric Banner", self.player))
-        set_rule(self.multiworld.get_entrance("Go To Lothric Castle", self.player),
-                 lambda state: state.has("Basin of Vows", self.player))
-        set_rule(self.multiworld.get_entrance("Go To Irithyll of the Boreal Valley", self.player),
-                 lambda state: state.has("Small Doll", self.player))
-        set_rule(self.multiworld.get_entrance("Go To Archdragon Peak", self.player),
-                 lambda state: state.has("Path of the Dragon", self.player))
-        set_rule(self.multiworld.get_entrance("Go To Grand Archives", self.player),
-                 lambda state: state.has("Grand Archives Key", self.player))
-        set_rule(self.multiworld.get_entrance("Go To Kiln of the First Flame", self.player),
-                 lambda state: state.has("Cinders of a Lord - Abyss Watcher", self.player) and
-                               state.has("Cinders of a Lord - Yhorm the Giant", self.player) and
-                               state.has("Cinders of a Lord - Aldrich", self.player) and
-                               state.has("Cinders of a Lord - Lothric Prince", self.player))
-
-        if self.multiworld.late_basin_of_vows[self.player] == Toggle.option_true:
-            add_rule(self.multiworld.get_entrance("Go To Lothric Castle", self.player),
-                     lambda state: state.has("Small Lothric Banner", self.player))
+        randomized_items = {item.name for item in self.local_itempool}
+
+        self._add_shop_rules()
+        self._add_npc_rules()
+        self._add_transposition_rules()
+        self._add_crow_rules()
+        self._add_allow_useful_location_rules()
+        self._add_early_item_rules(randomized_items)
+
+        self._add_entrance_rule("Firelink Shrine Bell Tower", "Tower Key")
+        self._add_entrance_rule("Undead Settlement", lambda state: (
+            state.has("Small Lothric Banner", self.player)
+            and self._can_get(state, "HWL: Soul of Boreal Valley Vordt")
+        ))
```

Comment:
When checking for a location, region or entrance from an entrance access condition, unless the spot in question is part of the source region of the entrance, you need to register an indirect condition.

I think Medic is on the case already

Reply:
You can use this:

https://github.com/NewSoupVi/Archipelago/tree/indirect_condition_checker

with this apworld & a DS3 yaml, to have it automatically detect missing indirect conditions

Reply:
I'm a little confused here. All the location checks for these entrance rules are exclusively coming from the source regions (HWL is the source for Undead Settlement, RS is the source for Cathedral of the Deep, and so on). But from https://github.com/nex3/Archipelago/pull/27, it sounds like this isn't the whole story. Is there some documentation I can read to understand this better? Do we need to worry about the *transitive conditions* for all locations/regions/entrances that can affect a given entrance rule?

Part of the reason we use so many `self._can_get` calls is to abstract out the transitive complexity, since different locations can depend on fairly complex conditions to become accessible. It's also possible if we end up doing something like entrance randomizer that these conditions will end up extremely contingent on particular RNG outcomes. Should we think about refactoring this to avoid checking location accessibility at all, and instead produce events for things like "boss defeated" so we can avoid this issue entirely?

Reply:
Idk how much you've been talking to Medic, they've been investigating the issue

I wasn't personally going to be able to confirm whether it's always the source region, but yes, you're right, if the entrance's source region and the location's parent region are the same, the indirect condition isn't necessary.

However, yes, you do need to also take care of transitive region requirements. The one case of this that actually tripped my indirect condition checker is this:

```
Dark Souls III, player 1: The parent region "Undead Settlement" of DarkSouls3Location "US: Soul of the Rotted Greatwood" could not be found in indirect_conditions for entrance "Go To Anor Londo" from "Irithyll of the Boreal Valley" to "Anor Londo". Suggested fix:
world.multiworld.register_indirect_condition(world.get_region("Undead Settlement"), world.get_entrance("Go To Anor Londo"))

This was caused by a recursive can_reach call.
The entrance access condition of "Go To Anor Londo" checked for
- DarkSouls3Location "IBV: Soul of Pontiff Sulyvahn", which checked for
- DarkSouls3Location "US: Soul of the Rotted Greatwood", which checked for
- the region in question, its parent region "Undead Settlement".
```

Now, it turns out that this is actually also fine, but in a non trivial way that my tool cannot account for. Medic explained to me that IBV can *only be reached through Undead Settlement*.

For an indirect condition to be necessary, two conditions have to be met:
- While inside an entrance access condition, a region is checked (This is transitive, so if an entrance checks a location, and that location checks another location, then that second location's region is also up for debate. I.e at some point before an entrance.can_reach has finished, somewhere down in the call stack, a region.can_reach is called)
- It's possible that the dependent region is reached "after" the source region of the entrance in a "random graph flood from menu region" kind of sense

Sorry for causing worry, I was not going to be able to confirm whether that was going to apply here or not, as I don't know anything about DS3.

If you want to read more about indirect conditions, the documentation is very lacking right now, but I have a PR open to remedy that somewhat: https://github.com/ArchipelagoMW/Archipelago/pull/3552
Although there I give a very "just make the indirect conditions" take, rather than the more nuanced "it might not always be necessary" version I just described to you (for public responsibility reasons)

Reply:
The explanation for why indirect conditions are necessary is this.

Region sweep is basically a Breadth First Search of the region graph from the Menu region, checking entrances one by one and adding newly reached nodes (regions) to the queue until there is nothing more to check.

However, if entrance access conditions depend on regions, then it is possible for an entrance to be checked and determined to be intraversible because the region hasn't been reached, then after that the region is reached by the graph search.

To remedy this, we would have to recheck all entrances every time a new region is reached, until no new regions are reached.

Because most games do not check for region access inside of entrance access conditions, AP has decided to **eschew this rechecking** and just checks every entrance once. This gives a significant performance gain to AP as a whole, about 30%-50%.

However, because some games *did* start using things like `region.can_reach` inside entrance access conditions, we provided a way to **manually** define that a *specific* entrance needs to be rechecked during region sweep if a *specific* region is reached during it. This keeps almost all of the performance upsides, even a game making heavy use of indirect conditions is still way way faster than if it just blanket "rechecks all entrances until nothing new is found".

However, we do recognize it's a beginner's trap and some games are very complex with their access rules.
There is a PR that makes this behavior optional via a world class attribute: https://github.com/ArchipelagoMW/Archipelago/pull/3682

If you have any further questions, I'm happy to answer them here, in #ap-world-dev, or in DMs on Discord :)

Reply:
Oh, to answer your last question:
Yes, if you have ER in the future the specific case in question with IBV and Undead Settlement could be a legitimate issue. The other cases are not gonna be an issue even with ER. If the source region of the entrance and the location's parent region are the same, it's 100% fine always.

Yes, events get around this issue, but they are **very slow**. If you can manage the indirect conditions, that'd be much much better.

Reply:
Sorry, it's a big info dump, but I got the vibe that you might be interested in the "long version" haha

Reply:
Thanks for the detailed background! It's very helpful to understand the underlying logic and trade-offs.

Is there a plan to pull the indirect condition checker into the main branch? I'm still a little nervous about missing conditions 😅

Reply:
Probably not. It's a hugely hacked together monkey patched version of AP. I would just use it with an .apworld or something, that's what I did in this case